### ISSUE TITLE
Hail Batch "multiple clients uploading"

### ISSUE THREAD

**Issue by tpoterba** [hail-is/hail#12950]
### What happened?

Daniel M is seeing this bug running the script [here](https://github.com/broadinstitute/tgg_methods/blob/master/vrs/vrs_annotation_batch.py#L275) with the flag --annotate-original.

The Zulip thread [here](https://hail.zulipchat.com/#narrow/stream/123010-Hail-Query-0.2E2-support/topic/Batch.20backend.20.22multiple.20clients.20uploading.22.3F/near/353337596) has more context as well.

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/batch/worker/worker.py", line 2150, in run
    await self.jvm.execute(local_jar_location, self.scratch, self.log_file, self.jar_url, self.argv)
  File "/usr/local/lib/python3.7/dist-packages/batch/worker/worker.py", line 2629, in execute
    raise JVMUserError(exception)
JVMUserError: java.util.concurrent.ExecutionException: java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
    at java.util.concurrent.FutureTask.report(FutureTask.java:122)
    at java.util.concurrent.FutureTask.get(FutureTask.java:192)
    at is.hail.JVMEntryway.retrieveException(JVMEntryway.java:224)
    at is.hail.JVMEntryway.finishFutures(JVMEntryway.java:186)
    at is.hail.JVMEntryway.main(JVMEntryway.java:156)
Caused by: java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
    at is.hail.JVMEntryway$1.run(JVMEntryway.java:107)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
    at java.lang.Thread.run(Thread.java:750)
Caused by: java.lang.reflect.InvocationTargetException
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:498)
    at is.hail.JVMEntryway$1.run(JVMEntryway.java:105)
    ... 7 more
Caused by: is.hail.relocated.com.google.cloud.storage.StorageException: Unable to recover in upload.
This may be a symptom of multiple clients uploading to the same upload session.

For debugging purposes:
uploadId: https://storage.googleapis.com/upload/storage/v1/b/rwalters-hail-tmp/o?name=merged_round2_sumstats.fix_lowconf.mt/entries/rows/parts/part-15801-2fde3786-67cb-42ed-8aac-f900cfcc4c00&uploadType=resumable&upload_id=ADPycduMEzX6d_uX4CiP6_XItJKmP8UnUnYBfyPoselMbyLUkxs1wDLPnxWl5gXr5LnBaVntYR_i7jchyxgVsRb_5PknvcCIcfDJ
chunkOffset: 16777216
chunkLength: 0
localOffset: 0
remoteOffset: 16777216
lastChunk: false


    at is.hail.relocated.com.google.cloud.storage.BlobWriteChannel.unrecoverableState(BlobWriteChannel.java:131)
    at is.hail.relocated.com.google.cloud.storage.BlobWriteChannel.unrecoverableState(BlobWriteChannel.java:87)
    at is.hail.relocated.com.google.cloud.storage.BlobWriteChannel.access$1000(BlobWriteChannel.java:35)
    at is.hail.relocated.com.google.cloud.storage.BlobWriteChannel$1.run(BlobWriteChannel.java:267)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
    at com.google.api.gax.retrying.DirectRetryingExecutor.submit(DirectRetryingExecutor.java:103)
    at is.hail.relocated.com.google.cloud.RetryHelper.run(RetryHelper.java:76)
    at is.hail.relocated.com.google.cloud.RetryHelper.runWithRetries(RetryHelper.java:50)
    at is.hail.relocated.com.google.cloud.storage.BlobWriteChannel.flushBuffer(BlobWriteChannel.java:189)
    at is.hail.relocated.com.google.cloud.BaseWriteChannel.flush(BaseWriteChannel.java:112)
    at is.hail.relocated.com.google.cloud.BaseWriteChannel.write(BaseWriteChannel.java:139)
    at is.hail.io.fs.GoogleStorageFS$$anon$2.flush(GoogleStorageFS.scala:270)
    at is.hail.io.fs.FSPositionedOutputStream.write(FS.scala:218)
    at java.io.DataOutputStream.write(DataOutputStream.java:107)
    at is.hail.utils.richUtils.ByteTrackingOutputStream.write(ByteTrackingOutputStream.scala:19)
    at is.hail.io.StreamBlockOutputBuffer.writeBlock(OutputBuffers.scala:293)
    at is.hail.io.LZ4OutputBlockBuffer.writeBlock(OutputBuffers.scala:313)
    at is.hail.io.BlockingOutputBuffer.writeBlock(OutputBuffers.scala:190)
    at is.hail.io.BlockingOutputBuffer.writeDouble(OutputBuffers.scala:234)
    at is.hail.io.LEB128OutputBuffer.writeDouble(OutputBuffers.scala:170)
    at __C9466collect_distributed_array_matrix_native_writer.__m13239ENCODE_SFloat64$_TO_o_float64(Unknown Source)
    at __C9466collect_distributed_array_matrix_native_writer.__m13243ENCODE_SBaseStructPointer_TO_o_struct_of_o_boolANDo_int32ANDo_float64ANDo_float64ANDo_float64ANDo_float64ANDo_float64END(Unknown Source)
    at __C9466collect_distributed_array_matrix_native_writer.__m13242ENCODE_SIndexablePointer_TO_r_array_of_o_struct_of_o_boolANDo_int32ANDo_float64ANDo_float64ANDo_float64ANDo_float64ANDo_float64END(Unknown Source)
    at __C9466collect_distributed_array_matrix_native_writer.__m13241ENCODE_SInsertFieldsStruct_TO_r_struct_of_r_array_of_o_struct_of_o_boolANDo_int32ANDo_float64ANDo_float64ANDo_float64ANDo_float64ANDo_float64ENDEND(Unknown Source)
    at __C9466collect_distributed_array_matrix_native_writer.__m9556split_Let(Unknown Source)
    at __C9466collect_distributed_array_matrix_native_writer.__m9536split_Let(Unknown Source)
    at __C9466collect_distributed_array_matrix_native_writer.apply(Unknown Source)
    at __C9466collect_distributed_array_matrix_native_writer.apply(Unknown Source)
    at is.hail.backend.BackendUtils.$anonfun$collectDArray$4(BackendUtils.scala:49)
    at is.hail.utils.package$.using(package.scala:635)
    at is.hail.annotations.RegionPool.scopedRegion(RegionPool.scala:162)
    at is.hail.backend.BackendUtils.$anonfun$collectDArray$3(BackendUtils.scala:48)
    at is.hail.backend.service.Worker$.$anonfun$main$10(Worker.scala:168)
    at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
    at is.hail.services.package$.retryTransientErrors(package.scala:119)
    at is.hail.backend.service.Worker$.main(Worker.scala:168)
    at is.hail.backend.service.Main$.main(Main.scala:32)
    at is.hail.backend.service.Main.main(Main.scala)
    ... 12 more

```

### Version

0.2.114

**Comment by danking** [comment]
The error message suggests that multiple clients are writing to:
```
gs://rwalters-hail-tmp/merged_round2_sumstats.fix_lowconf.mt/entries/rows/parts/part-15801-2fde3786-67cb-42ed-8aac-f900cfcc4c00
```
Which is a Hail Query partition file within a matrix table.

This is happening in a flush of GoogleStorageFS after we've filled up the local byte buffer. I don't think this is 0.2.114 because the line numbers don't line up. This appears to be before our recent changes to handle requester pays.

A couple thoughts:
1. We retryTransientErrors on the entire partition. What happens if a partition fails and we automatically retry it? Does the generated code clean up all the output streams when an exception occurs? If not, it's possible that there's still an open connection to GCS when we re-run the partition's generated code gain.
2. Hail Batch guarantees at least once execution of a job. It's possible that two distinct workers are executing the partition's generated code. Does the Hail Query generated code create a distinct file name for each execution? (I think the answer is yes.)

**Comment by danking** [comment]
Yeah, I think I agree that we have a unique UUID. What I'm more skeptical of is: what if we throw an exception in a `write`? Do we clean up all the open resources? If not, that could totally leave a writer open that will conflict when we retry (even if we retried on a different VM!).

---

`retryTransientErrors` expects partition code to be safe to execute twice, but otherwise it's quite simple:

```scala
  def retryTransientErrors[T](f: => T): T = {
    var delay = 0.1
    var errors = 0
    while (true) {
      try {
        return f
      } catch {
        case e: Exception =>
          errors += 1
          if (errors == 1 && isRetryOnceError(e))
            return f
          if (!isTransientError(e))
            throw e
          if (errors % 10 == 0)
            log.warn(s"encountered $errors transient errors, most recent one was $e")
      }
      delay = sleepAndBackoff(delay)
    }

    throw new AssertionError("unreachable")
  }
```
and this is the call site:
```scala
    val htc = new ServiceTaskContext(i)
    var result: Array[Byte] = null
    var userError: HailException = null
    try {
      retryTransientErrors {
        result = f(context, htc, theHailClassLoader, fs)
      }
    } catch {
      case err: HailException => userError = err
    }
    htc.close()
```

**Comment by tpoterba** [comment]
I think calling `f` multiple times will get different part files -- the UID uses java random, not seeded randomness.

**Comment by danking** [comment]
Agreed.

Do we insert finally blocks and clean up our open streams when an exception bubbles through generated code?

**Comment by danking** [comment]
Ah, sorry, you're totally right.

**Comment by tpoterba** [comment]
Darn, would have preferred to be wrong!

**Comment by danking** [comment]
I saw this again today in a fairly simple and isolated test. I'm beginning to wonder if this is just a new form of transient error. We pick 22 random characters from a 62 character alphabet. Odds of collision are minuscule:
```
In [2]: (1/62)**22
Out[2]: 3.693029961058969e-40
```
I verified `SecureRandom` with no constructor uses a randomly chosen seed.

There's three exceptions there (all the same one). The deepest one came during a write. The next two came during closes. The outermost exception is from the `using` cleaning up. I'm not sure where the middle exception comes from, I can't imagine who would try to `close` the stream other than `using`.

Regardless, it appears that the upload fails in some unrecoverable way. We're writing 2GiB in 256 8MiB chunks in this test, so we have more chances for something to go wrong. Maybe we just have to retry the entire partition when this happens?

https://ci.hail.is/batches/7404773/jobs/145
```
starting test is.hail.fs.gs.GoogleStorageFSSuite.testSeekMoreThanMaxInt...
Exception:
is.hail.relocated.com.google.cloud.storage.StorageException: Unable to recover in upload.
This may be a symptom of multiple clients uploading to the same upload session.

For debugging purposes:
uploadId: https://storage.googleapis.com/upload/storage/v1/b/hail-test-ezlis/o?name=fs-suite-tmp-6BO4gZ18Lheigp3ir9RSOh&uploadType=resumable&upload_id=ADPycduiXx2Jtiy_0Ll131_pPeEYKnnA23Hlk28_9TFESUMaubA9OqLK_n8Td5rPhTXnlpssGo796Q4bJxUeblhmSaYcCSWAMg2k
chunkOffset: 16777216
chunkLength: 8388608
localOffset: 1325400064
remoteOffset: 1342177280
lastChunk: false


	at is.hail.relocated.com.google.cloud.storage.BlobWriteChannel.unrecoverableState(BlobWriteChannel.java:131)
	at is.hail.relocated.com.google.cloud.storage.BlobWriteChannel.unrecoverableState(BlobWriteChannel.java:87)
	at is.hail.relocated.com.google.cloud.storage.BlobWriteChannel.access$1000(BlobWriteChannel.java:35)
	at is.hail.relocated.com.google.cloud.storage.BlobWriteChannel$1.run(BlobWriteChannel.java:267)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at com.google.api.gax.retrying.DirectRetryingExecutor.submit(DirectRetryingExecutor.java:103)
	at is.hail.relocated.com.google.cloud.RetryHelper.run(RetryHelper.java:76)
	at is.hail.relocated.com.google.cloud.RetryHelper.runWithRetries(RetryHelper.java:50)
	at is.hail.relocated.com.google.cloud.storage.BlobWriteChannel.flushBuffer(BlobWriteChannel.java:189)
	at is.hail.relocated.com.google.cloud.BaseWriteChannel.flush(BaseWriteChannel.java:112)
	at is.hail.relocated.com.google.cloud.BaseWriteChannel.write(BaseWriteChannel.java:139)
	at is.hail.io.fs.GoogleStorageFS$$anon$2.$anonfun$flush$1(GoogleStorageFS.scala:297)
	at is.hail.io.fs.GoogleStorageFS$$anon$2.doHandlingRequesterPays(GoogleStorageFS.scala:279)
	at is.hail.io.fs.GoogleStorageFS$$anon$2.flush(GoogleStorageFS.scala:297)
	at java.io.DataOutputStream.flush(DataOutputStream.java:123)
	at java.io.FilterOutputStream.close(FilterOutputStream.java:158)
	at is.hail.utils.package$.using(package.scala:640)
	at is.hail.fs.FSSuite.testSeekMoreThanMaxInt(FSSuite.scala:323)
	at is.hail.fs.FSSuite.testSeekMoreThanMaxInt$(FSSuite.scala:321)
	at is.hail.fs.gs.GoogleStorageFSSuite.testSeekMoreThanMaxInt(GoogleStorageFSSuite.scala:12)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:85)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:696)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:882)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1189)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:124)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:108)
	at org.testng.TestRunner.privateRun(TestRunner.java:767)
	at org.testng.TestRunner.run(TestRunner.java:617)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:348)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:343)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:305)
	at org.testng.SuiteRunner.run(SuiteRunner.java:254)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1224)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1149)
	at org.testng.TestNG.run(TestNG.java:1057)
	at org.testng.TestNG.privateMain(TestNG.java:1364)
	at org.testng.TestNG.main(TestNG.java:1333)
	Suppressed: is.hail.relocated.com.google.cloud.storage.StorageException: Unable to recover in upload.
This may be a symptom of multiple clients uploading to the same upload session.

For debugging purposes:
uploadId: https://storage.googleapis.com/upload/storage/v1/b/hail-test-ezlis/o?name=fs-suite-tmp-6BO4gZ18Lheigp3ir9RSOh&uploadType=resumable&upload_id=ADPycduiXx2Jtiy_0Ll131_pPeEYKnnA23Hlk28_9TFESUMaubA9OqLK_n8Td5rPhTXnlpssGo796Q4bJxUeblhmSaYcCSWAMg2k
chunkOffset: 16777216
chunkLength: 16777216
localOffset: 1325400064
remoteOffset: 1342177280
lastChunk: false


		at is.hail.relocated.com.google.cloud.storage.BlobWriteChannel.unrecoverableState(BlobWriteChannel.java:131)
		at is.hail.relocated.com.google.cloud.storage.BlobWriteChannel.unrecoverableState(BlobWriteChannel.java:87)
		at is.hail.relocated.com.google.cloud.storage.BlobWriteChannel.access$1000(BlobWriteChannel.java:35)
		at is.hail.relocated.com.google.cloud.storage.BlobWriteChannel$1.run(BlobWriteChannel.java:267)
		at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
		at com.google.api.gax.retrying.DirectRetryingExecutor.submit(DirectRetryingExecutor.java:103)
		at is.hail.relocated.com.google.cloud.RetryHelper.run(RetryHelper.java:76)
		at is.hail.relocated.com.google.cloud.RetryHelper.runWithRetries(RetryHelper.java:50)
		at is.hail.relocated.com.google.cloud.storage.BlobWriteChannel.flushBuffer(BlobWriteChannel.java:189)
		at is.hail.relocated.com.google.cloud.BaseWriteChannel.flush(BaseWriteChannel.java:112)
		at is.hail.relocated.com.google.cloud.BaseWriteChannel.write(BaseWriteChannel.java:139)
		at is.hail.io.fs.GoogleStorageFS$$anon$2.$anonfun$flush$1(GoogleStorageFS.scala:297)
		at is.hail.io.fs.GoogleStorageFS$$anon$2.doHandlingRequesterPays(GoogleStorageFS.scala:279)
		at is.hail.io.fs.GoogleStorageFS$$anon$2.flush(GoogleStorageFS.scala:297)
		at is.hail.io.fs.GoogleStorageFS$$anon$2.close(GoogleStorageFS.scala:306)
		at java.io.FilterOutputStream.close(FilterOutputStream.java:159)
		... 27 more
	Suppressed: is.hail.relocated.com.google.cloud.storage.StorageException: Unable to recover in upload.
This may be a symptom of multiple clients uploading to the same upload session.

For debugging purposes:
uploadId: https://storage.googleapis.com/upload/storage/v1/b/hail-test-ezlis/o?name=fs-suite-tmp-6BO4gZ18Lheigp3ir9RSOh&uploadType=resumable&upload_id=ADPycduiXx2Jtiy_0Ll131_pPeEYKnnA23Hlk28_9TFESUMaubA9OqLK_n8Td5rPhTXnlpssGo796Q4bJxUeblhmSaYcCSWAMg2k
chunkOffset: 16777216
chunkLength: 0
localOffset: 1325400064
remoteOffset: 1342177280
lastChunk: false


		at is.hail.relocated.com.google.cloud.storage.BlobWriteChannel.unrecoverableState(BlobWriteChannel.java:131)
		at is.hail.relocated.com.google.cloud.storage.BlobWriteChannel.unrecoverableState(BlobWriteChannel.java:87)
		at is.hail.relocated.com.google.cloud.storage.BlobWriteChannel.access$1000(BlobWriteChannel.java:35)
		at is.hail.relocated.com.google.cloud.storage.BlobWriteChannel$1.run(BlobWriteChannel.java:267)
		at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
		at com.google.api.gax.retrying.DirectRetryingExecutor.submit(DirectRetryingExecutor.java:103)
		at is.hail.relocated.com.google.cloud.RetryHelper.run(RetryHelper.java:76)
		at is.hail.relocated.com.google.cloud.RetryHelper.runWithRetries(RetryHelper.java:50)
		at is.hail.relocated.com.google.cloud.storage.BlobWriteChannel.flushBuffer(BlobWriteChannel.java:189)
		at is.hail.relocated.com.google.cloud.BaseWriteChannel.flush(BaseWriteChannel.java:112)
		at is.hail.relocated.com.google.cloud.BaseWriteChannel.write(BaseWriteChannel.java:139)
		at is.hail.io.fs.GoogleStorageFS$$anon$2.$anonfun$flush$1(GoogleStorageFS.scala:297)
		at is.hail.io.fs.GoogleStorageFS$$anon$2.doHandlingRequesterPays(GoogleStorageFS.scala:279)
		at is.hail.io.fs.GoogleStorageFS$$anon$2.flush(GoogleStorageFS.scala:297)
		at is.hail.io.fs.FSPositionedOutputStream.write(FS.scala:219)
		at java.io.DataOutputStream.write(DataOutputStream.java:107)
		at is.hail.fs.FSSuite.$anonfun$testSeekMoreThanMaxInt$1(FSSuite.scala:329)
		at is.hail.fs.FSSuite.$anonfun$testSeekMoreThanMaxInt$1$adapted(FSSuite.scala:323)
		at is.hail.utils.package$.using(package.scala:635)
		... 26 more
```

**Comment by jigold** [comment]
I did some digging into their client library. This exception (from the example where last_chunk=False) is thrown in the following cases:

```
  // Case 4: localNextByteOffset < remoteNextByteOffset
  //            && driftOffset > chunkSize:
  // Throw exception as remoteNextByteOffset has drifted beyond the retriable
  // chunk maintained in memory. This is not possible unless there's multiple
  // clients uploading to the same resumable upload session.

  // Case 8: remoteNextByteOffset==-1 && last == false && checkingForLastChunk
  // Not last chunk and checkingForLastChunk means this is the second time we
  // hit this case, meaning the upload was completed by a different client.

  // Case 9: Only possible if the client local offset continues beyond the remote
  // offset which is not possible.
```

I'm having a hard time following their code trying to trace how the `uploadId` is generated. Is it possible the issue is somehow the same `uploadId` is being used for multiple blob uploads because we're doing something that is not thread-safe??? One last thing I was trying to figure out but had a hard time following -- is it possible the requester pays additions changed the code path for how the `uploadId` is set if the problem has been more frequent after those changes?

**Comment by jigold** [comment]
If my hypothesis is true, then there should be evidence of API requests in the logs for different files having the same uploadId. I'll try and take a look at what the logs show for the failing file uploads.

**Comment by jigold** [comment]
Unfortunately, the logs aren't helpful as they only show completed requests.

**Comment by jigold** [comment]
Last thing before I give up for now. There's this comment in the BlobWriterRetry code:

```
/** Write channel implementation to upload Google Cloud Storage blobs. */
class BlobWriteChannel extends BaseWriteChannel<StorageOptions, BlobInfo> {

  private final ResultRetryAlgorithm<?> algorithmForWrite;
  // Detect if flushBuffer() is being retried or not.
  // TODO: I don't think this is thread safe, and there's probably a better way to detect a retry
  // occuring.
  private boolean retrying = false;
  private boolean checkingForLastChunk = false;
```

**Comment by danking** [comment]
Observed again  https://batch.hail.is/batches/7942284/jobs/203

This method isn't threaded at all: https://github.com/hail-is/hail/blob/5d545c8d9fb474ea211e350c5dc48c599db6f5ea/hail/src/test/scala/is/hail/fs/FSSuite.scala#L339-L368

The temporary location is drawn as 62 choose 22. So, odds of collision are 3 * 10^16.

I can't find the referenced case analysis in Google's latest code. [It is present in this fork](https://github.com/leogamas/java-storage/blob/2af8dfd95cdebc9e4d8252b0bbe3f092844d9f2c/google-cloud-storage/src/main/java/com/google/cloud/storage/BlobWriteChannel.java#L68-L198) from a few years ago. There seems to have been a rewrite [two months ago](https://github.com/googleapis/java-storage/blame/main/google-cloud-storage/src/main/java/com/google/cloud/storage/BlobWriteChannel.java) (here's [the main commit](https://github.com/googleapis/java-storage/commit/1b52a1053130620011515060787bada10c324c0b)). That landed in [2.25.0](https://github.com/googleapis/java-storage/releases/tag/v2.25.0) which was released in July.

```
is.hail.relocated.com.google.cloud.storage.StorageException: Unable to recover in upload.
This may be a symptom of multiple clients uploading to the same upload session.

For debugging purposes:
uploadId: https://storage.googleapis.com/upload/storage/v1/b/hail-test-ezlis/o?name=fs-suite-tmp-2LzGioRNy6RqIS2pfXIoSO&uploadType=resumable&upload_id=ADPycdvZ5HhnGfOKt5TE1qXWiHpqIpZnXVTYWuWUCXNPRF9HqyCB-4LvRsxNX6SUWRgk13pYrzYaa9-wXlvNZt1oct0ptaEz0bS3
chunkOffset: 16777216
chunkLength: 8388608
localOffset: 268435456
remoteOffset: 285212672
lastChunk: false


	at is.hail.relocated.com.google.cloud.storage.BlobWriteChannel.unrecoverableState(BlobWriteChannel.java:131)
	at is.hail.relocated.com.google.cloud.storage.BlobWriteChannel.unrecoverableState(BlobWriteChannel.java:87)
	at is.hail.relocated.com.google.cloud.storage.BlobWriteChannel.access$1000(BlobWriteChannel.java:35)
	at is.hail.relocated.com.google.cloud.storage.BlobWriteChannel$1.run(BlobWriteChannel.java:267)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at com.google.api.gax.retrying.DirectRetryingExecutor.submit(DirectRetryingExecutor.java:103)
	at is.hail.relocated.com.google.cloud.RetryHelper.run(RetryHelper.java:76)
	at is.hail.relocated.com.google.cloud.RetryHelper.runWithRetries(RetryHelper.java:50)
	at is.hail.relocated.com.google.cloud.storage.BlobWriteChannel.flushBuffer(BlobWriteChannel.java:189)
	at is.hail.relocated.com.google.cloud.BaseWriteChannel.flush(BaseWriteChannel.java:112)
	at is.hail.relocated.com.google.cloud.BaseWriteChannel.write(BaseWriteChannel.java:139)
	at is.hail.io.fs.GoogleStorageFS$$anon$2.$anonfun$flush$1(GoogleStorageFS.scala:317)
	at is.hail.io.fs.GoogleStorageFS$$anon$2.doHandlingRequesterPays(GoogleStorageFS.scala:299)
	at is.hail.io.fs.GoogleStorageFS$$anon$2.flush(GoogleStorageFS.scala:317)
	at java.io.DataOutputStream.flush(DataOutputStream.java:123)
	at java.io.FilterOutputStream.close(FilterOutputStream.java:158)
	at is.hail.utils.package$.using(package.scala:640)
	at is.hail.fs.FSSuite.testSeekMoreThanMaxInt(FSSuite.scala:341)
	at is.hail.fs.FSSuite.testSeekMoreThanMaxInt$(FSSuite.scala:339)
	at is.hail.fs.gs.GoogleStorageFSSuite.testSeekMoreThanMaxInt(GoogleStorageFSSuite.scala:12)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:85)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:696)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:882)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1189)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:124)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:108)
	at org.testng.TestRunner.privateRun(TestRunner.java:767)
	at org.testng.TestRunner.run(TestRunner.java:617)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:348)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:343)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:305)
	at org.testng.SuiteRunner.run(SuiteRunner.java:254)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1224)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1149)
	at org.testng.TestNG.run(TestNG.java:1057)
	at org.testng.TestNG.privateMain(TestNG.java:1364)
	at org.testng.TestNG.main(TestNG.java:1333)
	Suppressed: is.hail.relocated.com.google.cloud.storage.StorageException: Unable to recover in upload.
This may be a symptom of multiple clients uploading to the same upload session.

For debugging purposes:
uploadId: https://storage.googleapis.com/upload/storage/v1/b/hail-test-ezlis/o?name=fs-suite-tmp-2LzGioRNy6RqIS2pfXIoSO&uploadType=resumable&upload_id=ADPycdvZ5HhnGfOKt5TE1qXWiHpqIpZnXVTYWuWUCXNPRF9HqyCB-4LvRsxNX6SUWRgk13pYrzYaa9-wXlvNZt1oct0ptaEz0bS3
chunkOffset: 16777216
chunkLength: 16777216
localOffset: 268435456
remoteOffset: 285212672
lastChunk: false


		at is.hail.relocated.com.google.cloud.storage.BlobWriteChannel.unrecoverableState(BlobWriteChannel.java:131)
		at is.hail.relocated.com.google.cloud.storage.BlobWriteChannel.unrecoverableState(BlobWriteChannel.java:87)
		at is.hail.relocated.com.google.cloud.storage.BlobWriteChannel.access$1000(BlobWriteChannel.java:35)
		at is.hail.relocated.com.google.cloud.storage.BlobWriteChannel$1.run(BlobWriteChannel.java:267)
		at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
		at com.google.api.gax.retrying.DirectRetryingExecutor.submit(DirectRetryingExecutor.java:103)
		at is.hail.relocated.com.google.cloud.RetryHelper.run(RetryHelper.java:76)
		at is.hail.relocated.com.google.cloud.RetryHelper.runWithRetries(RetryHelper.java:50)
		at is.hail.relocated.com.google.cloud.storage.BlobWriteChannel.flushBuffer(BlobWriteChannel.java:189)
		at is.hail.relocated.com.google.cloud.BaseWriteChannel.flush(BaseWriteChannel.java:112)
		at is.hail.relocated.com.google.cloud.BaseWriteChannel.write(BaseWriteChannel.java:139)
		at is.hail.io.fs.GoogleStorageFS$$anon$2.$anonfun$flush$1(GoogleStorageFS.scala:317)
		at is.hail.io.fs.GoogleStorageFS$$anon$2.doHandlingRequesterPays(GoogleStorageFS.scala:299)
		at is.hail.io.fs.GoogleStorageFS$$anon$2.flush(GoogleStorageFS.scala:317)
		at is.hail.io.fs.GoogleStorageFS$$anon$2.close(GoogleStorageFS.scala:326)
		at java.io.FilterOutputStream.close(FilterOutputStream.java:159)
		... 27 more
	Suppressed: is.hail.relocated.com.google.cloud.storage.StorageException: Unable to recover in upload.
This may be a symptom of multiple clients uploading to the same upload session.

For debugging purposes:
uploadId: https://storage.googleapis.com/upload/storage/v1/b/hail-test-ezlis/o?name=fs-suite-tmp-2LzGioRNy6RqIS2pfXIoSO&uploadType=resumable&upload_id=ADPycdvZ5HhnGfOKt5TE1qXWiHpqIpZnXVTYWuWUCXNPRF9HqyCB-4LvRsxNX6SUWRgk13pYrzYaa9-wXlvNZt1oct0ptaEz0bS3
chunkOffset: 16777216
chunkLength: 0
localOffset: 268435456
remoteOffset: 285212672
lastChunk: false


		at is.hail.relocated.com.google.cloud.storage.BlobWriteChannel.unrecoverableState(BlobWriteChannel.java:131)
		at is.hail.relocated.com.google.cloud.storage.BlobWriteChannel.unrecoverableState(BlobWriteChannel.java:87)
		at is.hail.relocated.com.google.cloud.storage.BlobWriteChannel.access$1000(BlobWriteChannel.java:35)
		at is.hail.relocated.com.google.cloud.storage.BlobWriteChannel$1.run(BlobWriteChannel.java:267)
		at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
		at com.google.api.gax.retrying.DirectRetryingExecutor.submit(DirectRetryingExecutor.java:103)
		at is.hail.relocated.com.google.cloud.RetryHelper.run(RetryHelper.java:76)
		at is.hail.relocated.com.google.cloud.RetryHelper.runWithRetries(RetryHelper.java:50)
		at is.hail.relocated.com.google.cloud.storage.BlobWriteChannel.flushBuffer(BlobWriteChannel.java:189)
		at is.hail.relocated.com.google.cloud.BaseWriteChannel.flush(BaseWriteChannel.java:112)
		at is.hail.relocated.com.google.cloud.BaseWriteChannel.write(BaseWriteChannel.java:139)
		at is.hail.io.fs.GoogleStorageFS$$anon$2.$anonfun$flush$1(GoogleStorageFS.scala:317)
		at is.hail.io.fs.GoogleStorageFS$$anon$2.doHandlingRequesterPays(GoogleStorageFS.scala:299)
		at is.hail.io.fs.GoogleStorageFS$$anon$2.flush(GoogleStorageFS.scala:317)
		at is.hail.io.fs.FSPositionedOutputStream.write(FS.scala:227)
		at java.io.DataOutputStream.write(DataOutputStream.java:107)
		at is.hail.fs.FSSuite.$anonfun$testSeekMoreThanMaxInt$1(FSSuite.scala:347)
		at is.hail.fs.FSSuite.$anonfun$testSeekMoreThanMaxInt$1$adapted(FSSuite.scala:341)
		at is.hail.utils.package$.using(package.scala:635)
		... 26 more

test is.hail.fs.gs.GoogleStorageFSSuite.testSeekMoreThanMaxInt FAILURE
```

**Comment by danking** [comment]
I haven't observed this since we merged that change (though it introduced a different issue, https://github.com/hail-is/hail/issues/13721 🤦 ).
